backend/models/user.models.ts

```diff
@@ -0,0 +1,62 @@
+import {model, Schema} from 'mongoose';
+import bcrypt = require('bcryptjs');
+import jwt = require('jsonwebtoken');
+import { databaseSecret } from '../environment';
+
+interface User {
+  id: number;
+  username: string;
+  password: string;
+  tokens: [];
+}
+
+// User Schema
+export const UserSchema = new Schema({
+  username: {
+    type: String,
+    required: true,
+    unique: true
+  },
+  password: {
+    type: String,
+    required: true,
+    minLength: 6
+  },
+  tokens: [{
+    token: {
+      type: String,
+      required: true
+    }
+  }]
+});
+
+UserSchema.pre('save', async function(next) {
+  const user = this;
+  if (user.isModified('password')) {
+    user.password = await bcrypt.hash(user.password, 8);
+  }
+  next();
+});
+
+UserSchema.methods.generateAuthToken = async function() {
+  const user = this;
+  const token = jwt.sign({_id: user._id}, databaseSecret);
```

Comment:
is `databaseSecret` some sort of password/encryption for both user and server sides?

Reply:
Secret key. It allows the application to communicate with the database. Otherwise anyone could connect to it. (More for production environments than anything)